internal/iapl/policy.go

```diff
@@ -257,20 +361,183 @@ func (v *policy) expandActionBindings() {
+}
+
+// createRoleBindingResourceType creates a role-binding resource type contains
+// a list of all the actions.
+// The role-binding resources will be used to create a 3-way relationship
+// between a resource, a subject and a role
+func (v *policy) createRoleBindingResourceType() {
+	rolebinding := ResourceType{
+		Name:     v.p.RBAC.RoleBindingResource.Name,
+		IDPrefix: v.p.RBAC.RoleBindingResource.IDPrefix,
+	}
+
+	role := Relationship{
+		Relation: RolebindingRoleRelation,
+		TargetTypes: []types.TargetType{
+			{Name: v.p.RBAC.RoleResource.Name},
+		},
+	}
+
+	// 2. create relationship to subjects
+	subjects := Relationship{
+		Relation:    RolebindingSubjectRelation,
+		TargetTypes: v.p.RBAC.RoleBindingSubjects,
+	}
+
+	// 3. create a list of action-bindings representing permissions for all the
+	// actions in the policy
+	actionbindings := make([]ActionBinding, 0, len(v.ac))
+
+	for actionName := range v.ac {
+		ab := ActionBinding{
+			ActionName: actionName,
+			TypeName:   v.p.RBAC.RoleBindingResource.Name,
+			ConditionSets: []types.ConditionSet{
+				{
+					Conditions: []types.Condition{
+						{
+							RelationshipAction: &types.ConditionRelationshipAction{
+								Relation:   RolebindingRoleRelation,
+								ActionName: actionName + PermissionRelationSuffix,
+							},
+						},
+					},
+				},
+				{
+					Conditions: []types.Condition{
+						{RelationshipAction: &types.ConditionRelationshipAction{Relation: RolebindingSubjectRelation}},
+					},
+				},
+			},
+		}
+
+		actionbindings = append(actionbindings, ab)
+	}
+
+	v.bn = append(v.bn, actionbindings...)
+
+	// 4. create role-binding resource type
+	rolebinding.Relationships = []Relationship{role, subjects}
+	v.rt[v.p.RBAC.RoleBindingResource.Name] = rolebinding
+}
+
+// expandRBACV2Relationships adds RBAC V2 relationships to all the resource
+// types that has `ResourceRoleBindingV2` defined.
+// Relationships like member_roles, available_roles, are created to support
+// role inheritance, e.g., an org should be able to use roles defined by its
+// parents
+func (v *policy) expandRBACV2Relationships() {
+	for name, resourceType := range v.rt {
+		// not all roles are available for all resources, available roles are
+		// the roles that a resource owners (if it is a role-owner) or inherited
+		// from their owner or parent
+		availableRoles := []Condition{}
+
+		// if resource type is a role-owner, add the role-relationship to the
+		// resource
+		if _, ok := v.RBAC().RoleOwnersSet()[name]; ok {
+			memberRoleRelation := Relationship{
+				Relation: RoleOwnerMemberRoleRelation,
+				TargetTypes: []types.TargetType{
+					{Name: v.p.RBAC.RoleResource.Name},
+				},
+			}
+
+			resourceType.Relationships = append(resourceType.Relationships, memberRoleRelation)
+
+			// i.e. avail_role = member_role
+			availableRoles = append(availableRoles, Condition{
+				RelationshipAction: &ConditionRelationshipAction{
+					Relation: RoleOwnerMemberRoleRelation,
```

Comment:
Is this meant to say that a role is available if a relationship exists from the owner to the role?

Reply:
yes. combined with the following code it constructs a concept of
"the roles that are available to you, as a resource, are the roles that are available to your owner or parent, plus the roles that you own (if you're a role owner)"